packages/block-editor/src/components/colors/use-colors.js

```diff
@@ -89,23 +95,34 @@ export default function __experimentalUseColors(
 	const createComponent = useMemo(
 		() =>
 			memoize(
-				( property, color, colorValue, customColor ) => ( { children } ) =>
+				( name, property, className, color, colorValue, customColor ) => ( {
+					children,
+					className: componentClass = '',
+					style: componentStyles = {},
+				} ) =>
 					// Clone children, setting the style property from the color configuration,
 					// if not already set explicitly through props.
 					Children.map( children, ( child ) => {
-						let className = child.props.className;
-						let style = child.props.style;
+						const elementClassName = classnames(
+							componentClass,
+							child.props.className,
+							{
+								[ `has-${ kebabCase( color ) }-${ kebabCase( property ) }` ]: color,
+								[ className || `has-${ kebabCase( name ) }` ]: color || customColor,
+							}
+						);
+						let style = {
+							...componentStyles,
+							...( child.props.style || {} ),
+						};
+
 						if ( color ) {
-							className = `${ child.props.className } has-${ kebabCase(
-								color
-							) }-${ kebabCase( property ) }`;
-							style = { [ property ]: colorValue, ...child.props.style };
+							style = { [ property ]: colorValue, ...style };
 						} else if ( customColor ) {
-							className = `${ child.props.className } has-${ kebabCase( property ) }`;
-							style = { [ property ]: customColor, ...child.props.style };
+							style = { [ property ]: customColor, ...style };
 						}
```

Comment:
This can be reduced to a ternary now.

Reply:
We would need something like:
```
						let style = {
							...( color ?
								( { [ property ]: colorValue } ) :
								( customColor ? { [ property ]: customColor } : {} )
							),
							...componentStyle,
							...( child.props.style || {} ),
						};
```

And we have a lint rule against nested ternaries.

Reply:
No, we could just do:

```js
let style = {
	...( ( color || customColor ) && {
		[ property ]: color ? colorValue : customColor,
	} ),
	...componentStyle,
	...child.props.style,
};
```